File: incre-tests/synduce/constraints/sortedlist/parallel_max2.f
Inductive List = elt Int | cons {Int, List};
Inductive CList = single Int | concat {CList, CList};

cat = fix (
  \f: List -> List -> List. \a: List. \b: List.
  match a with
    elt x -> cons {x, b}
  | cons {h, t} -> cons {h, (f t b)}
  end
);

repr = fix (
  \f: CList -> List. \xs: CList.
  match xs with
    single x -> elt x
  | concat {a, b} -> cat (f a) (f b)
  end
);

max = \a: Int. \b: Int. if < a b then b else a;
lmax = fix (
  \f: CList -> Int. \xs: CList.
  match xs with
    single x -> x
  | concat {x, y} -> max (f x) (f y)
  end
);

min = \a: Int. \b: Int. if < a b then a else b;
lmin = fix (
  \f: CList -> Int. \xs: CList.
  match xs with
    single x -> x
  | concat {x, y} -> min (f x) (f y)
  end
);

is_part = fix (
  \f: CList -> Bool. \c: CList.
  match c with
    single w -> true
  | concat {x, y} ->
    and (> (lmin x) (lmax y)) (and (f x) (f y))
  end
);

spec = fix (
  \f: List -> Int. \xs: List.
  match xs with
    elt x -> x
  | cons {h, t} -> max h (f t)
  end
);

target = fix (
  \f: CList -> Compress CList. \c: CList.
  match c with
    single w -> c
  | concat {x, y} -> concat {f x, y} /*Avoid the recursion of y*/
  end
);

main = \xs: CList. if is_part xs then spec (repr (target xs)) else 0;

File: ml-benchmark/synduce/constraints/sortedlist/parallel_max2.ml
type list = Elt of int | Cons of int * list
type clist = Single of int | Concat of clist * clist

let rec cat a b =
  match a with
  | Elt x -> Cons (x, b)
  | Cons (h, t) -> Cons (h, cat t b)

let rec repr xs =
  match xs with
  | Single x -> Elt x
  | Concat (a, b) -> cat (repr a) (repr b)

let max a b = if a < b then b else a

let rec lmax xs =
  match xs with
  | Single x -> x
  | Concat (x, y) -> max (lmax x) (lmax y)

let min a b = if a < b then a else b

let rec lmin xs =
  match xs with
  | Single x -> x
  | Concat (x, y) -> min (lmin x) (lmin y)

let rec is_part c =
  match c with
  | Single _ -> true
  | Concat (x, y) -> (lmin x) > (lmax y) && is_part x && is_part y

let rec spec xs =
  match xs with
  | Elt x -> x
  | Cons (h, t) -> max h (spec t)

let rec target c =
  match c with
  | Single _ -> c
  | Concat (x, y) -> Concat (target x, y)

let program xs =
  if is_part xs then spec (repr (target xs)) else 0